kvserver,storage: add ExpectExclusionSince to Get

stevendanna opened the pull request:
This adds ExpectExclusionSince to Get requests. This option semantically works the same as the ExpectExclusionSince option on Put and Del requests. The implementation, however, is slightly different. It adjusts the read timestamp used by storage.MVCCGet and then configures the scanner to return an ExclusionViolationError rather than a WriteTooOld error.

Informs #142977

Release note: None

stevendanna committed:
kvserver,storage: add ExpectExclusionSince to Get

This adds ExpectExclusionSince to Get requests. This option semantically works
the same as the ExpectExclusionSince option on Put and Del requests. The
implementation, however, is slightly different. It adjusts the read timestamp
used by storage.MVCCGet and then configures the scanner to return an
ExclusionViolationError rather than a WriteTooOld error.

Informs #142977

Release note: None

arulajmani reviewed (commented):
Mostly looks good; haven't looked at the testing yet.

> Reviewed 4 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-:-OSyuqAV7X4IN3QLbHBz:b-nqt5be)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @xxmplus)

___
*[`pkg/kv/kvserver/batcheval/cmd_get.go` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSyt1zM4VJ2Lg5gUu9A:-OSyt1zM4VJ2Lg5gUu9B:b-safnyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4812cd21fc63afecd1b52db7a04099d84204fb91/pkg/kv/kvserver/batcheval/cmd_get.go#L21)):*
> ```Go
> 
> func init() {
> 	RegisterReadWriteCommand(kvpb.Get, DefaultDeclareIsolatedKeys, Get)
> ```

Do we also need to change the latching logic here? In particular, do we need to ensure that in cases where this flag is set, we acquire latches from `req.ExpectExclusionSince` instead of `header.Timestamp`? 
___
*[`pkg/kv/kvpb/api.proto` line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSys4Ch24huESKpl5xE:-OSys4Ch24huESKpl5xF:b-8b3yf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4812cd21fc63afecd1b52db7a04099d84204fb91/pkg/kv/kvpb/api.proto#L229)):*
> ```proto
>   // ExpectExclusionSince, if set, indicates that the request should return an
>   // error if this key has been written to at a timestamp at or after the given
>   // timestamp. This is different from the ReadTimestamp of the transaction in
> ```

nit: consider rephrasing this as "It may be used by a transaction to verify an unreplicated lock acquired ..." while also including the caveat that we're not checking for the existence of a lock, as the lock may have been lost, but just that the protection it would have provided still holds.
___
*[`pkg/kv/kvpb/api.proto` line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSysSM-7HfECS0QvIhD:-OSysSM-7HfECS0QvIhE:bmx7qos) ([raw file](https://github.com/cockroachdb/cockroach/blob/4812cd21fc63afecd1b52db7a04099d84204fb91/pkg/kv/kvpb/api.proto#L230)):*
> ```proto
>   // error if this key has been written to at a timestamp at or after the given
>   // timestamp. This is different from the ReadTimestamp of the transaction in
>   // that it allows a writer to verify an unreplicated lock acquired at an
> ```

nit: "allows a writer" seems off. 
___
*[`pkg/storage/mvcc.go` line 1243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSyuRhmDf6sOI-96p5t:-OSyuRhmDf6sOI-96p5u:b-vn5oqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4812cd21fc63afecd1b52db7a04099d84204fb91/pkg/storage/mvcc.go#L1243)):*
> ```Go
> 	// if a more recent version of a key is found.
> 	WriteTooOldErrorOnMoreRecent
> 	// ExclusionViolationErrorOnMoreRecent indicates that MVCCGet should return an ExclusionLostError
> ```

Instead of pushing this into mvcc.go, could we instead transform a `WriteTooOldError` in `Get` (`cmd_get.go`) to an `ExclusionViolationError` instead?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148364)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @xxmplus)

___
*[`pkg/kv/kvserver/batcheval/cmd_get.go` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSyt1zM4VJ2Lg5gUu9A:-OSyz6bT1kj-iCAqRF5M:byrfjr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4812cd21fc63afecd1b52db7a04099d84204fb91/pkg/kv/kvserver/batcheval/cmd_get.go#L21)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we also need to change the latching logic here? In particular, do we need to ensure that in cases where this flag is set, we acquire latches from `req.ExpectExclusionSince` instead of `header.Timestamp`?
</blockquote></details>

Hrmmmmmm.  I don't _think_ so. Here is my argument

 For non-locking requests, we don't allow this. If we did, we would be taking a SpanRead latch. Since ExpectExclusionSince <= header.Timestamp, taking the latch at the earlier timestamp would _increase_ the writers that could be concurrent with it so I think the more conservative approach would be to still take it at header.Timestamp.

 For locking requests, we have two cases:

- Shared locks acquire a SpanRead latch but at MaxTimestamp so it is effectively isolated from all other writers. It is possible other locking readers will acquire shared locks. But I think that is OK since all this option every tells us is whether there was a read between ExpectExclusionSince and header.Timestamp.

- Exclusive locks acquire SpanReadWrite latch and will be isolated from all other writers

Let me know if something there smells wrong to you.


<!-- Sent from Reviewable.io -->

stevendanna committed:
kvserver,storage: handle ExpectExclusionSince during eval

Epic: none
Release note: None

stevendanna committed:
kvserver: allow ExclusionViolationError in SSI transactions

Get will return ExclusionViolationError instead of WriteTooOld errors.
While we could check the IsoLevel before doing the transformation, I
can't see a good reason not to return the ExclusionViolationError
error. The ExclusionViolationError won't force a refresh, but that
makes sense since the refresh will fail almost by definition.

Epic: none
Release note: None

stevendanna committed:
kvpb: update comment on ExpectExclusionSince

Epic: none
Release note: None

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148364)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @xxmplus)

___
*[`pkg/kv/kvpb/api.proto` line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSys4Ch24huESKpl5xE:-OSzOA5T2ZXmNz_9bMmB:b-jg510o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc6ec664418cd73ba2411bc9b5a84a12ae989a/pkg/kv/kvpb/api.proto#L229)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider rephrasing this as "It may be used by a transaction to verify an unreplicated lock acquired ..." while also including the caveat that we're not checking for the existence of a lock, as the lock may have been lost, but just that the protection it would have provided still holds.
</blockquote></details>

Updated the comment, if you like it, I'll make a similar change on the other requests.
___
*[`pkg/storage/mvcc.go` line 1243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSyuRhmDf6sOI-96p5t:-OSz2qXR3N90RoZkntkd:b-i14moi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc6ec664418cd73ba2411bc9b5a84a12ae989a/pkg/storage/mvcc.go#L1243)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Instead of pushing this into mvcc.go, could we instead transform a `WriteTooOldError` in `Get` (`cmd_get.go`) to an `ExclusionViolationError` instead?
</blockquote></details>

:D Thanks for the nudge. I thought of this as well but only after all this code had been written and was working and then didn't come back to this.  


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
TFTR! Github is having some issues so my response has been delivered in pieces.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-:-OSzWd6YEFcNmB8XpZOZ:bplp4tt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @xxmplus)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo bit about capturing the latching discussion in a comment somewhere. 

> Reviewed 7 of 11 files at r1, 6 of 6 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-:-OT7aJrr3z_DmOrlcNSp:b-jcej0z)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @stevendanna and @xxmplus)

___
*[`pkg/kv/kvpb/api.proto` line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSys4Ch24huESKpl5xE:-OT7aFsoB4C027JcM7rx:b6rdql7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc6ec664418cd73ba2411bc9b5a84a12ae989a/pkg/kv/kvpb/api.proto#L229)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Updated the comment, if you like it, I'll make a similar change on the other requests.
</blockquote></details>

Yup, looks good to me. 
___
*[`pkg/kv/kvserver/batcheval/cmd_get.go` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSyt1zM4VJ2Lg5gUu9A:-OT7XU7W6d_R9gfcv2O1:btm3eu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc6ec664418cd73ba2411bc9b5a84a12ae989a/pkg/kv/kvserver/batcheval/cmd_get.go#L21)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Hrmmmmmm.  I don't _think_ so. Here is my argument

 For non-locking requests, we don't allow this. If we did, we would be taking a SpanRead latch. Since ExpectExclusionSince <= header.Timestamp, taking the latch at the earlier timestamp would _increase_ the writers that could be concurrent with it so I think the more conservative approach would be to still take it at header.Timestamp.

 For locking requests, we have two cases:

- Shared locks acquire a SpanRead latch but at MaxTimestamp so it is effectively isolated from all other writers. It is possible other locking readers will acquire shared locks. But I think that is OK since all this option every tells us is whether there was a read between ExpectExclusionSince and header.Timestamp.

- Exclusive locks acquire SpanReadWrite latch and will be isolated from all other writers

Let me know if something there smells wrong to you.
</blockquote></details>

No, all this checks out to me. I confused myself a bit here when I wrote the comment. 

Could you capture these latching considerations in a comment somewhere? Maybe right where the field is added to the request? 
___
*[`pkg/storage/mvcc.go` line 1243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148364#-OSyuRhmDf6sOI-96p5t:-OT7_mE37LWRT8fmEfyw:b-qr9y7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc6ec664418cd73ba2411bc9b5a84a12ae989a/pkg/storage/mvcc.go#L1243)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

:D Thanks for the nudge. I thought of this as well but only after all this code had been written and was working and then didn't come back to this.
</blockquote></details>

Nice, this looks slick now.


<!-- Sent from Reviewable.io -->
